### PR TITLE
#90759 EventReader: use appropriate message lock value

### PR DESCRIPTION
**Pull request by tabman83**

### CODE REVIEW

File: src/CaptainHook.Common/ServiceBus/ServiceBusManager.cs [hunk 1]
```diff
@@ -74,7 +75,7 @@ public async Task DeleteSubscriptionAsync(string topicName, string subscriptionN
             {
                 await topic.Subscriptions
                     .Define(subscriptionName.ToLowerInvariant())
-                    .WithMessageLockDurationInSeconds(60)
+                    .WithMessageLockDurationInSeconds(messageLockDurationInSeconds)
```
Comment (chomick) [inline src/CaptainHook.Common/ServiceBus/ServiceBusManager.cs:78]: not sure if you have another task for it but the story asks for this time to be updated when reader is refreshed as well

File: src/CaptainHook.EventReaderService/EventReaderService.cs [hunk 2]
```diff
@@ -194,11 +201,25 @@ protected override async Task OnChangeRoleAsync(ReplicaRole newRole, Cancellatio
         /// </summary>
         internal int InFlightMessageCount => HandlerCount - _freeHandlers.Count;
 
+        internal int CalculateMessageLock()
+        {
```
Comment (chomick) [inline src/CaptainHook.EventReaderService/EventReaderService.cs:205]: is the fact that we select a random settings, correct? what if i define different timeouts and dueations for different endpoibts? the original plan was to have it in subscriber level (webhook classl but if you guys decided to di it on ebdpoint, how will you handle different settings here?

File: src/CaptainHook.EventReaderService/EventReaderService.cs [hunk 3]
```diff
@@ -194,11 +201,25 @@ protected override async Task OnChangeRoleAsync(ReplicaRole newRole, Cancellatio
...
+            var routesDurations = _initData.SubscriberConfiguration.WebhookRequestRules
+                                .FirstOrDefault(x => x.Routes.Any())?
+                                .Routes?
+                                .Select(x => x.RetrySleepDurations);
+                
+            var durations = routesDurations ?? new List<TimeSpan[]> { _initData.SubscriberConfiguration.RetrySleepDurations };
+
+            return durations.Max(d => _messageLockDurationCalculator.CalculateAsSeconds(_initData.SubscriberConfiguration.Timeout, d));
```
Comment (chomick) [inline src/CaptainHook.EventReaderService/EventReaderService.cs:213]: was the timeout not checked on endpoint level in handler? is that correct value here?

File: src/CaptainHook.Common/ServiceBus/ServiceBusManager.cs (hunk 1 above)
Comment (tabman83) [reply src/CaptainHook.Common/ServiceBus/ServiceBusManager.cs:78]: What is a reader refresh? We only have destroy & re-create reader hence why this is placed here.

Comment (chomick) [reply src/CaptainHook.Common/ServiceBus/ServiceBusManager.cs:78]: But when you create new reader, the subscription is already in place - will it update the lock duration settings?

File: src/CaptainHook.EventReaderService/EventReaderService.cs (hunk 2 above)
Comment (chomick) [reply src/CaptainHook.EventReaderService/EventReaderService.cs:205]: I can see now that you are getting a list of retry-duration, not a first one from the list - this should be better.

File: src/CaptainHook.EventReaderService/EventReaderService.cs (hunk 3 above)
Comment (chomick) [reply src/CaptainHook.EventReaderService/EventReaderService.cs:213]: I see now that you use Max() from calculated duration - that should be fine.

File: src/CaptainHook.Common/ServiceBus/ServiceBusManager.cs (hunk 1 above)
Comment (tabman83) [reply src/CaptainHook.Common/ServiceBus/ServiceBusManager.cs:86]: Yes thanks. This has now been addressed.

**Review by chomick — APPROVED** [review]
Nino, I think we should test it properly on CI or Test before we move to higher env's. Should we attempt to create more e2e tests related to this functionality? Those tests could use API now for setup/tear-down and they can verify (broadly) number of retries, timeouts etc.
It would take a bit of work to write them up but I think this is well spent effort - what do you think?

**Comment by tabman83** [comment]
> Nino, I think we should test it properly on CI or Test before we move to higher env's. Should we attempt to create more e2e tests related to this functionality? Those tests could use API now for setup/tear-down and they can verify (broadly) number of retries, timeouts etc.
> It would take a bit of work to write them up but I think this is well spent effort - what do you think?

I'll devise a plan and share with you

**Review by nikhilgirraj — APPROVED** [review]

**Pull request merged**